### PR TITLE
Report bug `SA_FIELD_SELF_ASSIGNMENT` in nested classes as well

### PR DESCRIPTION
**Pull request by baloghadamsoftware**
Bug `SA_FIELD_SELF_ASSIGNMENT` was not reported in nested classes, only in the outer class. This lead to inconsistent behavior. See issue ([#2142](https://github.com/spotbugs/spotbugs/issues/2142)). This PR fixes this issue.

----

Make sure these boxes are checked before submitting your PR -- thank you!

- [X] Added an entry into `CHANGELOG.md` if you have changed SpotBugs code

### CODE REVIEW

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java [hunk 1]
```diff
@@ -91,8 +93,12 @@ public void sawOpcode(int seen) {
                 OpcodeStack.Item fourth = stack.getStackItem(3);
                 XField f2 = third.getXField();
                 int registerNumber2 = fourth.getRegisterNumber();
-                if (f2 != null && f2.equals(getXFieldOperand()) && registerNumber2 >= 0
-                        && registerNumber2 == third.getFieldLoadedFromRegister()
+                int loadedFromRegister2 = fourth.getFieldLoadedFromRegister();
+                if (f2 != null && f2.equals(getXFieldOperand())
+                        && ((registerNumber2 >= 0 && registerNumber2 == third.getFieldLoadedFromRegister())
+                                || (loadedFromRegister2 >= 0 && third.getUserValue() != null
+                                        && third.getUserValue() instanceof Integer
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java:100]: Is instance is an `Integer`, we're sure that it is not-null.

```suggestion
                                || (loadedFromRegister2 >= 0 && third.getUserValue() instanceof Integer
```

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java [hunk 2]
```diff
@@ -91,8 +93,12 @@ public void sawOpcode(int seen) {
...
+                                        && loadedFromRegister2 == (Integer) third.getUserValue()))
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java:101]: Better to do not depends on auto boxing.
```suggestion
                                        && loadedFromRegister2 == ((Integer) third.getUserValue()).intValue()))
```

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java [hunk 3]
```diff
@@ -161,7 +167,28 @@ public void sawOpcode(int seen) {
...
+                parentInstanceLoadFromRegister = top.getFieldLoadedFromRegister();
+            }
+        }
     }
 
+    @Override
+    public void afterOpcode(int seen) {
+        super.afterOpcode(seen);
+
+        if (seen == Const.GETFIELD && parentInstanceLoadFromRegister > -1) {
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java:188]: Could you replace this magic number `-1` with a `static final int` field?
Better to tell our intention to who is reading our code.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java [hunk 4]
```diff
@@ -161,7 +167,28 @@ public void sawOpcode(int seen) {
...
+            OpcodeStack.Item top = stack.getStackItem(0);
+            top.setUserValue(parentInstanceLoadFromRegister);
```
Comment (KengoTODA) [inline spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindFieldSelfAssignment.java:190]: when you use `UserValue` in your detector, please annotate the detector with `@CustomUserValue`.

refs https://blog.kengo-toda.jp/entry/2013/10/13/011552

**Review by KengoTODA — CHANGES_REQUESTED** [review]

**Comment by baloghadamsoftware** [comment]
Thank you for your comments, @KengoTODA! I think I fixed every one of them now.

**Review by KengoTODA — APPROVED** [review]
LGTM, thank you!

**Review by ThrawnCA — DISMISSED** [review]

**Review by KengoTODA — APPROVED** [review]

**Pull request merged**